### PR TITLE
webm/theora/yuv2rgb/libsimplewebm: Fix colour issues I could find.

### PR DESCRIPTION
**Pull request by 20kdc**
Modifies yuv2rgb to remove endian-dependency in one of the functions, correctly orders RGBA, and deduplicates the macros.

The Theora and WEBM video players, rather than using reordered UV inputs to fix the yuv2rgb issue (which produces some distortion), now use the correct YUV order, since yuv2rgb is outputting correctly.

Modifies libsimplewebm to output RGBA so it can support the full range of RGBA formats in future.
libsimplewebm is now dependent on yuv2rgb as a result.

NOTE: This is by no means a complete solution for all WEBM formats.
The library is still limited to planar formats.

The following test project contains the WEBM and Theora files I made for testing.
They display correctly in VLC, so I am assuming they are correct.
[VideoColourLab.zip](https://github.com/godotengine/godot/files/2879771/VideoColourLab.zip)

It'd be a good idea to test this on more WEBM files before continuing.

Additional note: The side effects shown when U/V gets swapped suggests issue 18208 *may* be caused by this, but don't rely on it.

Here's the screenshots from v3.1-beta4, Arch Linux x86\_64:
image
image
(To my knowledge, master looks the same. This PR shows them all correctly, but of course that's my own test-cases, so please do test with more WEBMs.)

### CODE REVIEW

**Comment by groud** [comment]
We usually do not modify thirparty libraries, and when we do, it is via patches or carefully delimited area.
I would suggest you to propose the changes upstream first, and see if they could be merged there instead.

**Comment by akien-mga** [comment]
CC @zaps166

**Comment by 20kdc** [comment]
Ah, thank you. I'm a bit worried about upstream because the changes needed for handling the various different VP9 modes (including, say, the RGBA modes) essentially mean that libsimplewebm has to output RGBA for consistency between all the different flavours of VP9, which of course breaks API compatibility. It also means yuv2rgb becomes a dependency of libsimplewebm, at least the way I've set things up; I feel it'd be better if yuv2rgb was treated as shared code between both to reduce duplication, but that has other potential problems...
An alternate route would be for me to undo the RGBA pipeline stuff, so libsimplewebm isn't changed as much or at all.
This said in turn, yuv2rgb is *also* altered by these changes. I do not know how pushing that upstream is going to work...

File: modules/webm/video_stream_webm.cpp [hunk 1]
```diff
@@ -307,40 +305,19 @@ void VideoStreamPlaybackWebm::update(float p_delta) {
...
-							// 								converted = true;
-						}
+				{
+					// Scope isolation just in case Write is picky about when it occurs.
+					PoolVector<uint8_t>::Write w = frame_data.write();
+					image.w = webm->getWidth();
+					image.h = webm->getHeight();
+					image.data = w.ptr();
+					err = video->getImage(image);
+				}
```
Comment (LikeLakers2) [inline modules/webm/video_stream_webm.cpp]: Out of curiousity, what's with the extra set of `{ }` brackets in this section? The `if` statement just above this already has its own `{` opening bracket.

Comment (LikeLakers2) [reply modules/webm/video_stream_webm.cpp]: Wait nevermind, I misread the comment included at the beginning of this section. Woops!

**Review by zaps166 — COMMENTED** [review]
I don't thing that simple webm library should convert the image. Almost every video uses YUV planar format, so it should be converted later, before displaying the image (especially on GPU e.g. via GLSL).

How fast is applying alpha? Isn't it slow for large video frames (depends on what compiler do with this)?

File: thirdparty/libsimplewebm/VPXDecoder.hpp [hunk 2]
```diff
@@ -69,12 +63,13 @@ class VPXDecoder
...
 
 	bool decode(const WebMFrame &frame);
-	IMAGE_ERROR getImage(Image &image); //The data is NOT copied! Only 3-plane, 8-bit images are supported.
+	IMAGE_ERROR getImage(Image &image); // Data is copied to a user-owned buffer. See Image.
 
 private:
 	vpx_codec_ctx *m_ctx;
 	const void *m_iter;
 	int m_delay;
+	int last_space;
```
[the remark this answers is not in the document]
Comment (20kdc) [reply thirdparty/libsimplewebm/VPXDecoder.hpp]: The 'use YUV so GLSL can be later used' makes sense, but that means libsimplewebm will have to expose the RGB flag to Godot. Should I go with this? (It sounds like a solution that'll keep API compatibility, at least.)
If so, I'll PR directly to libsimplewebm.
Also, as for applying alpha, it only matters if it's a format that *supports* alpha, and I've been having trouble making a WEBM file of that kind in the first place, so if I go through with this I'm dropping alpha support from the plan under untestability grounds.

**Comment by 20kdc** [comment]
With upstream now containing the necessary feature, should I have a separate PR for the thirdparty update, or should I perform it as part of this one?

**Comment by akien-mga** [comment]
You can use this PR directly by rebasing it. Maybe make a first commit with only the sync from upstream (also update the hash documented in `thirdparty/README.md`), and a second commit with the remaining changes.

**Comment by 20kdc** [comment]
It appears 'OpusVorbisDecoder::getPCMF' is a function that for some reason has been removed or was not contributed to libsimplewebm...?

**Comment by zaps166** [comment]
@20kdc 
IIRC `OpusVorbisDecoder::getPCMF` was not contributed to libsimplewebm.

**Comment by 20kdc** [comment]
...Ok, so... getting the resulting diff shows there's a lot of unmarked changes. Biggest is that it looks like the `libwebm` usage within libsimplewebm was updated in the Godot version.
There's the addition of `std::no_throw` in various places, as well.
```
--- a/libwebm/README.libvpx
+++ b/libwebm/README.libvpx
@@ -1,5 +1,5 @@
 URL: https://chromium.googlesource.com/webm/libwebm
-Version: 32d5ac49414a8914ec1e1f285f3f927c6e8ec29d
+Version: d7c62173ff6b4a5e0a2f86683a5b67db98cf09bf
 License: BSD
 License File: LICENSE.txt
```

**Comment by 20kdc** [comment]
The current state of this is that I didn't sync the libwebm part, as it appears to have been updated separately by e71f1099 and 7c0da764.
The main contents of libsimplewebm have been synced with Godot-local changes marked.

**Comment by 20kdc** [comment]
...force-pushed again in line with the clang-format checker.

**Comment by 20kdc** [comment]
So, er... should I do another rebase to keep up, or... I've lost track of what next to do here.

**Comment by guelinator** [comment]
@akien-mga can you tell me if this pull request will be merged into the master branch or not.
in my opinion the vp8/theora implementation currently in godot 3.1 is worthless, because colors are not reproduced correctly. as far as i can see, the patch from 20kdc successfully fixes the colors, and thats the way it should be.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**